conversion/conversion.go

```diff
@@ -433,38 +438,70 @@ func UpdateDDLForeignKeys(project, instance, dbName string, conv *internal.Conv,
+	msg := fmt.Sprintf("Updating schema of database %s in instance %s with foreign key constraints ...\n", dbName, instance)
 	p := internal.NewProgress(int64(len(fkStmts)), msg, internal.Verbose())
-	for i, fkStmt := range fkStmts {
-		// TODO: Improve performance of the foreign key constraint updates.
-		// For example, issue all of the update ops first before waiting for them to complete
-		// so that that can execute in parallel. We could also print out ids of the
-		// long-running operations.
-		op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
-			Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
-			Statements: []string{fkStmt},
-		})
-		if err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
-			continue
-		}
-		if err := op.Wait(ctx); err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+
+	type Result struct {
+		fkStmt string
+		err    error
+	}
+	workers := make(chan int, MaxWorkers)
+	results := make(chan Result, len(fkStmts))
+
+	for i := 1; i <= MaxWorkers; i++ {
+		workers <- i
+	}
+
+	// We dispatch parallel foreign key create requests to ensure the backfill runs in parallel to reduce overall time.
+	// This cuts down the time taken to a third (approx) compared to Serial and Batched creation. We also do not want to create
+	// too many requests and get throttled due to network or hitting catalog memory limits.
+	// Ensure atmost `MaxWorkers` go routines run in parallel that each update the ddl with one foreign key statement.
+	for _, fkStmt := range fkStmts {
+		workerId := <-workers
+		fmt.Printf("worker %d handling %s\n", workerId, fkStmt)
+		go func(fkStmt string, workerId int) {
+			defer func() { workers <- workerId }()
+			fmt.Printf("Submitting new FK create request for %s: %s\n", dbName, fkStmt)
+			op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
+				Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
+				Statements: []string{fkStmt},
+			})
+			if err != nil {
+				fmt.Printf("Cannot submit request for create foreign key with statement %s: %s\n", fkStmt, err)
+				conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+				results <- Result{fkStmt, err}
+				return
+			}
+			if err := op.Wait(ctx); err != nil {
+				fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
+				conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+				results <- Result{fkStmt, err}
+				return
+			}
+			internal.VerbosePrintln("Updated schema with statement: " + fkStmt)
+			results <- Result{fkStmt, nil}
+		}(fkStmt, workerId)
+	}
+	progress := int64(0)
```

Comment:
We will get here only when <= workers fkStmt are left to be processed -- that will make the end user feel like foreign key updates are stuck when looking at the progress. For example, if there are 100 fkStmts and 10 workers, then users will start seeing progress after 90 fkStmts are already processed (the progress will go very fast from 1 to 90 in that case). 

Thus, I wonder if it makes sense to increment the progress in the defer statement above? (the next follow up question in that case would be whether we really need results array and Result struct?

Reply:
Good catch!!